compiler/rustc_passes/src/custom_abi.rs

```diff
@@ -0,0 +1,153 @@
+use rustc_abi::ExternAbi;
+use rustc_hir::def::DefKind;
+use rustc_hir::def_id::{LocalDefId, LocalModDefId};
+use rustc_hir::intravisit::Visitor;
+use rustc_hir::{self as hir, ExprKind, FnSig};
+use rustc_middle::hir::nested_filter::OnlyBodies;
+use rustc_middle::query::Providers;
+use rustc_middle::ty::{self, TyCtxt};
+use rustc_span::{BytePos, sym};
+
+use crate::errors::{
+    AbiCustomCall, AbiCustomClothedFunction, AbiCustomSafeForeignFunction, AbiCustomSafeFunction,
+};
+
+pub(crate) fn provide(providers: &mut Providers) {
+    *providers = Providers { check_mod_custom_abi, ..*providers };
+}
+
+fn check_mod_custom_abi(tcx: TyCtxt<'_>, module_def_id: LocalModDefId) {
+    if !tcx.features().abi_custom() {
+        return;
+    }
+
+    let items = tcx.hir_module_items(module_def_id);
+    for def_id in items.definitions() {
+        let def_kind = tcx.def_kind(def_id);
+
+        // An `extern "custom"` function cannot be marked as `safe`.
+        if let DefKind::ForeignMod = def_kind
+            && let hir::Node::Item(item) = tcx.hir_node_by_def_id(def_id)
+            && let hir::ItemKind::ForeignMod { abi: ExternAbi::Custom, items } = item.kind
+        {
+            for item in items {
+                let hir_id = item.id.hir_id();
+                if let hir::Node::ForeignItem(foreign_item) = tcx.hir_node(hir_id)
+                    && let hir::ForeignItemKind::Fn(sig, _, _) = foreign_item.kind
+                    && sig.header.is_safe()
+                {
+                    let len = "safe ".len() as u32;
+                    let safe_span = sig.span.shrink_to_lo().with_hi(sig.span.lo() + BytePos(len));
```

Comment:
This feels somewhat brittle. `rustc_ast::ast::Safety` has the span for this, if you can get that here. Or perhaps this entire pass should be in `rustc_ast_passes` instead, similar to the `MissingAbi` lint/errors?

Reply:
If the decision is to reject function params and return types (TBD at https://github.com/rust-lang/rust/issues/140566#issuecomment-2860606615, not sure what your preference is), that could probably also be part of the AST pass.

Reply:
it is fragile, but kind of needed for good suggestions. The `rustc_ast::ast::Safety` span is just the keyword, it's missing the space that also must be deleted, and there is no good way of getting the span of what follows as far as I can see.

Moving this logic over to `rustc_ast_passes` does work nicely.